Issues with django.forms.ModelChoiceField and ModelMultipleChoiceField

Hello, I have been investigating the feasibility of publishing a set of Hypothesis strategies and helpers for testing [Wagtail](https://github.com/wagtail/wagtail) projects. Wagtail is a CMS Library for Python, built on Django. I've been pleased with the results I've had using Hypothesis to test some of my other projects, and think there would be some benefit to integrating property-based testing into Wagtail projects (I am eager to find ways to improve the quality of software delivered to clients).

While working on this, I've come across some issues with the strategies that are generated for Django's `ModelChoiceField`, which Wagtail uses frequently, and `ModelMultipleChoiceField`.

The following test cases illustrate the issues.

```python
import hypothesis
import pytest

from django import forms
from django.contrib.auth.models import User
from hypothesis import strategies as st
from hypothesis.extra.django import TestCase, from_field, from_form


class UserSelectForm(forms.Form):
    user = forms.ModelChoiceField(User.objects.all())
    users = forms.ModelMultipleChoiceField(User.objects.all())


@pytest.mark.django_db()
class TestModelChoicesField(TestCase):
    @classmethod
    def setUpTestData(cls) -> None:
        User.objects.create_user(
            username="foo-user", email="foo@user.com", password="not-so-secure"
        )
        User.objects.create_user(
            username="bar-user", email="bar@user.com", password="not-so-secure"
        )

    @hypothesis.given(
        # Must use deferred or get error for database access
        user=st.deferred(lambda: from_field(forms.ModelChoiceField(User.objects.all())))
    )
    def test_user(self, user):
        assert isinstance(
            user, forms.models.ModelChoiceIteratorValue
        )  # Expect a User instance
        assert isinstance(user.value, int)  # Inner value is an int (primary key)

    @hypothesis.given(
        # Must use deferred or get error for database access
        users=st.deferred(
            lambda: from_field(forms.ModelMultipleChoiceField(User.objects.all()))
        )
    )
    def test_users(self, users):
        assert isinstance(
            users, forms.models.ModelChoiceIteratorValue
        )  # Expect QuerySet[User]
        assert isinstance(users.value, int)

    # Deferred not required
    @hypothesis.given(form=from_form(UserSelectForm))
    def test_form(self, form):
        assert isinstance(form.data["user"], forms.models.ModelChoiceIteratorValue)
        assert not form.is_valid()
        assert form.errors == {
            "user": [
                "Select a valid choice. That choice is not one of the available choices."
            ],
            "users": ["Enter a list of values."],
        }
```

### 1. `from_field` with `ModelChoiceField` requires using a `deferred` strategy

This makes sense, as it looks like `from_field` generates a `sampled_from(field.choices)` strategy, which causes database access to evaluate the choices.

```
$ pytest -k 'ModelChoices and test_user' --reuse-db
======================================================= test session starts =======================================================platform linux -- Python 3.11.9, pytest-8.1.1, pluggy-1.5.0
django: version: 5.0.6, settings: wagtail_hypothesis.test.settings (from ini)
rootdir: /home/jmunn/projects/wagtail-hypothesis
configfile: pyproject.toml
plugins: xdist-3.6.1, hypothesis-6.103.0, cov-5.0.0, django-4.8.0
collected 27 items / 1 error / 27 deselected / 0 selected                                                                         

============================================================= ERRORS ==============================================================_______________________________________ ERROR collecting tests/test_model_choices_field.py ________________________________________tests/test_model_choices_field.py:16: in <module>
    class TestModelChoicesField(TestCase):
tests/test_model_choices_field.py:28: in TestModelChoicesField
    user=from_field(forms.ModelChoiceField(User.objects.all()))
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/hypothesis/extra/django/_fields.py:304: in from_field
    for value, name_or_optgroup in field.choices:
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/forms/models.py:1422: in __iter__
    for obj in queryset:
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/db/models/query.py:518: in _iterator
    yield from iterable
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/db/models/query.py:91: in __iter__
    results = compiler.execute_sql(
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/db/models/sql/compiler.py:1558: in execute_sql
    cursor = self.connection.chunked_cursor()
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/db/backends/base/base.py:670: in chunked_cursor
    return self.cursor()
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/utils/asyncio.py:26: in inner
    return func(*args, **kwargs)
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/db/backends/base/base.py:316: in cursor
    return self._cursor()
../../.virtualenvs/wagtail-hypothesis-dev/lib/python3.11/site-packages/django/db/backends/base/base.py:292: in _cursor
    self.ensure_connection()
E   RuntimeError: Database access not allowed, use the "django_db" mark, or the "db" or "transactional_db" fixtures to enable it.
===================================================== short test summary info =====================================================ERROR tests/test_model_choices_field.py - RuntimeError: Database access not allowed, use the "django_db" mark, or the "db" or "transactional_db" fixtures to enable it.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!================================================= 27 deselected, 1 error in 0.37s =================================================
```

A similar error is reported when using django's `manage.py test` test runner instead of pytest. This is a little confusing as the behaviour differs from using `from_form` in a `given` decorator, where it seems the field strategies are not eagerly evaluated. From my perspective, it would be ideal if `from_field` handled delaying the database access to the time that the first example is evaluated, which seems to be in a context where the database connection is available. The consistency between different field types would make this more user friendly.

### 2. Generated strategies yield incorrect types

The strategies generated for both `ModelChoiceField` and `ModelMultipleChoiceField` yield instances of `forms.models.ModelChoiceIteratorValue`, which are not valid values for a form — see the final test. I would expect either a model instance or a queryset, respectively, to be generated (or a primary key or list of primary keys — I need to double-check the order of validation and upcasting in the form instantiation and validation pathways). My expectation is that `from_form` should generate values that are valid for the defined form fields — is that assumption correct?

---

I am happy to submit a patch for these issues.

--- a/AUTHORS.rst
+++ b/AUTHORS.rst
@@ -96,6 +96,7 @@ their individual contributions.
 * `Jonathan Gayvallet <https://github.com/Meallia>`_ (jonathan.gayvallet@orange.com)
 * `Jonty Wareing <https://www.github.com/Jonty>`_ (jonty@jonty.co.uk)
 * `Joshua Boone <https://www.github.com/patchedwork>`_ (joshuaboone4190@gmail.com)
+* `Joshua Munn <https://www.github.com/jams2>`_ (public@elysee-munn.family)
 * `jmhsi <https://www.github.com/jmhsi>`_
 * `Justus Magin <https://github.com/keewis>`_
 * `jwg4 <https://www.github.com/jwg4>`_

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,6 @@
+RELEASE_TYPE: minor
+
+This release adds strategies for Django's ``ModelChoiceField`` and
+``ModelMultipleChoiceField`` (:issue:`4010`).
+
+Thanks to Joshua Munn for this contribution.

--- a/hypothesis-python/src/hypothesis/extra/django/_fields.py
+++ b/hypothesis-python/src/hypothesis/extra/django/_fields.py
@@ -262,8 +262,7 @@ def _for_form_boolean(field):
     return st.booleans()
 
 
-@register_for(df.ModelChoiceField)
-def _for_model_choice(field):
+def _model_choice_strategy(field):
     def _strategy():
         if field.choices is None:
             # The field was instantiated with queryset=None, and not
@@ -293,7 +292,19 @@ def _for_model_choice(field):
             ]
         )
 
+    # Accessing field.choices causes database access, so defer the strategy.
     return st.deferred(_strategy)
+
+
+@register_for(df.ModelChoiceField)
+def _for_model_choice(field):
+    return _model_choice_strategy(field)
+
+
+@register_for(df.ModelMultipleChoiceField)
+def _for_model_multiple_choice(field):
+    min_size = 1 if field.required else 0
+    return st.sets(_model_choice_strategy(field), min_size=min_size).map(list)
 
 
 def register_field_strategy(

--- a/hypothesis-python/src/hypothesis/extra/django/_fields.py
+++ b/hypothesis-python/src/hypothesis/extra/django/_fields.py
@@ -344,6 +344,12 @@ def from_field(field: F) -> st.SearchStrategy[Union[F, None]]:
     instance attributes such as string length and validators.
     """
     check_type((dm.Field, df.Field), field, "field")
+
+    # The following isinstance check must occur *before* the getattr
+    # check. In the case of ModelChoicesField, evaluating
+    # field.choices causes database access, which we want to avoid if
+    # we don't have a connection (the generated strategies for
+    # ModelChoicesField defer evaluation of `choices').
     if not isinstance(field, df.ModelChoiceField) and getattr(field, "choices", False):
         choices: list = []
         for value, name_or_optgroup in field.choices:

--- a/hypothesis-python/src/hypothesis/extra/django/_fields.py
+++ b/hypothesis-python/src/hypothesis/extra/django/_fields.py
@@ -304,7 +304,7 @@ def _for_model_choice(field):
 @register_for(df.ModelMultipleChoiceField)
 def _for_model_multiple_choice(field):
     min_size = 1 if field.required else 0
-    return st.sets(_model_choice_strategy(field), min_size=min_size).map(list)
+    return st.lists(_model_choice_strategy(field), min_size=min_size, unique=True)
 
 
 def register_field_strategy(

--- a/hypothesis-python/src/hypothesis/extra/django/_fields.py
+++ b/hypothesis-python/src/hypothesis/extra/django/_fields.py
@@ -262,6 +262,40 @@ def _for_form_boolean(field):
     return st.booleans()
 
 
+@register_for(df.ModelChoiceField)
+def _for_model_choice(field):
+    def _strategy():
+        if field.choices is None:
+            # The field was instantiated with queryset=None, and not
+            # subsequently updated.
+            raise InvalidArgument(
+                "Cannot create strategy for ModelChoicesField with no choices"
+            )
+        elif hasattr(field, "_choices"):
+            # The choices property was set manually.
+            choices = field._choices
+        else:
+            # choices is not None, and was not set manually, so we
+            # must have a QuerySet.
+            choices = field.queryset
+
+        if not choices.ordered:
+            raise InvalidArgument("QuerySet must have explicit ordering")
+
+        return st.sampled_from(
+            [
+                (
+                    choice.value
+                    if isinstance(choice, df.models.ModelChoiceIteratorValue)
+                    else choice  # Empty value, if included.
+                )
+                for choice, _ in field.choices
+            ]
+        )
+
+    return st.deferred(_strategy)
+
+
 def register_field_strategy(
     field_type: Type[AnyField], strategy: st.SearchStrategy
 ) -> None:
@@ -299,7 +333,7 @@ def from_field(field: F) -> st.SearchStrategy[Union[F, None]]:
     instance attributes such as string length and validators.
     """
     check_type((dm.Field, df.Field), field, "field")
-    if getattr(field, "choices", False):
+    if not isinstance(field, df.ModelChoiceField) and getattr(field, "choices", False):
         choices: list = []
         for value, name_or_optgroup in field.choices:
             if isinstance(name_or_optgroup, (list, tuple)):

--- a/hypothesis-python/src/hypothesis/extra/django/_fields.py
+++ b/hypothesis-python/src/hypothesis/extra/django/_fields.py
@@ -279,7 +279,11 @@ def _model_choice_strategy(field):
             choices = field.queryset
 
         if not choices.ordered:
-            raise InvalidArgument("QuerySet must have explicit ordering")
+            raise InvalidArgument(
+                f"Cannot create strategy for {field.__class__.__name__} with a choices "
+                "attribute derived from a QuerySet without an explicit ordering - this may "
+                "cause Hypothesis to produce unstable results between runs."
+            )
 
         return st.sampled_from(
             [

--- a/hypothesis-python/tests/django/toystore/forms.py
+++ b/hypothesis-python/tests/django/toystore/forms.py
@@ -19,11 +19,13 @@ from django.core.validators import (
 from django.forms import widgets
 
 from tests.django.toystore.models import (
+    Company,
     CouldBeCharming,
     Customer,
     ManyNumerics,
     ManyTimes,
     OddFields,
+    Store,
 )
 
 
@@ -221,3 +223,11 @@ class UsernameForm(ReprForm):
 
 class ReadOnlyPasswordHashFieldForm(ReprForm):
     password = ReadOnlyPasswordHashField()
+
+
+class StoreForm(ReprModelForm):
+    company = forms.ModelChoiceField(queryset=Company.objects.order_by("name"))
+
+    class Meta:
+        model = Store
+        fields = "__all__"

--- a/hypothesis-python/tests/django/toystore/forms.py
+++ b/hypothesis-python/tests/django/toystore/forms.py
@@ -231,3 +231,9 @@ class StoreForm(ReprModelForm):
     class Meta:
         model = Store
         fields = "__all__"
+
+
+class MultipleCompaniesForm(ReprForm):
+    companies = forms.ModelMultipleChoiceField(
+        queryset=Company.objects.order_by("name")
+    )

--- a/hypothesis-python/tests/django/toystore/test_given_forms.py
+++ b/hypothesis-python/tests/django/toystore/test_given_forms.py
@@ -30,6 +30,7 @@ from tests.django.toystore.forms import (
     ManyMultiValueForm,
     ManyNumericsForm,
     ManyTimesForm,
+    MultipleCompaniesForm,
     OddFieldsForm,
     RegexFieldForm,
     ShortStringForm,
@@ -175,6 +176,20 @@ class TestFormsWithModelChoices(TestCase):
         self.assertEqual(choice, "")
 
     @given(form=from_form(StoreForm))
-    def test_valid_store(self, form):
+    def test_store_form_valid(self, form):
         assume(form.data["company"])
+        self.assertTrue(form.is_valid())
+
+    @given(
+        choice=from_field(
+            forms.ModelMultipleChoiceField(queryset=Company.objects.order_by("name"))
+        )
+    )
+    def test_from_model_multiple_choices_field(self, choice):
+        n_choices = len(choice)
+        self.assertEqual(n_choices, len(set(choice)))
+        self.assertEqual(n_choices, Company.objects.filter(pk__in=choice).count())
+
+    @given(form=from_form(MultipleCompaniesForm))
+    def test_multiple_companies_form_valid(self, form):
         self.assertTrue(form.is_valid())

--- a/hypothesis-python/tests/django/toystore/test_given_forms.py
+++ b/hypothesis-python/tests/django/toystore/test_given_forms.py
@@ -8,8 +8,15 @@
 # v. 2.0. If a copy of the MPL was not distributed with this file, You can
 # obtain one at https://mozilla.org/MPL/2.0/.
 
-from hypothesis import given
-from hypothesis.extra.django import TestCase, from_form, register_field_strategy
+from django import forms
+
+from hypothesis import assume, given
+from hypothesis.extra.django import (
+    TestCase,
+    from_field,
+    from_form,
+    register_field_strategy,
+)
 from hypothesis.strategies import booleans, sampled_from
 
 from tests.django.toystore.forms import (
@@ -27,12 +34,14 @@ from tests.django.toystore.forms import (
     RegexFieldForm,
     ShortStringForm,
     SlugFieldForm,
+    StoreForm,
     TemporalFieldForm,
     URLFieldForm,
     UsernameForm,
     UUIDFieldForm,
     WithValidatorsForm,
 )
+from tests.django.toystore.models import Company
 
 register_field_strategy(
     BroadBooleanField, booleans() | sampled_from(["1", "0", "True", "False"])
@@ -127,3 +136,45 @@ class TestGetsBasicForms(TestCase):
     @given(from_form(UsernameForm))
     def test_read_only_password_hash_field_form(self, password_form):
         self.assertTrue(password_form.is_valid())
+
+
+class TestFormsWithModelChoices(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        # Set up example Company records to use as choices for
+        # Store.company. These must exist before creating a strategy
+        # for the ModelChoiceField.
+        cls.company_names = ("Bill's Flowers", "Jane's Sporting Goods")
+        for name in cls.company_names:
+            Company.objects.create(name=name)
+
+    @given(
+        choice=from_field(
+            forms.ModelChoiceField(queryset=Company.objects.order_by("name"))
+        )
+    )
+    def test_from_model_choices_field(self, choice):
+        assume(choice != "")  # Skip the empty choice.
+        self.assertIsInstance(choice, int)
+        Company.objects.get(id=choice)
+
+    @given(
+        choice=from_field(
+            forms.ModelChoiceField(
+                queryset=Company.objects.order_by("name"), empty_label=None
+            )
+        )
+    )
+    def test_from_model_choices_field_no_empty_choice(self, choice):
+        Company.objects.get(id=choice)
+
+    @given(choice=from_field(forms.ModelChoiceField(queryset=Company.objects.none())))
+    def test_from_model_choices_field_empty(self, choice):
+        self.assertEqual(choice, "")
+
+    @given(form=from_form(StoreForm))
+    def test_valid_store(self, form):
+        assume(form.data["company"])
+        self.assertTrue(form.is_valid())

--- a/hypothesis-python/tox.ini
+++ b/hypothesis-python/tox.ini
@@ -160,19 +160,19 @@ commands =
 commands =
     pip install .[pytz]
     pip install django~=3.2.15
-    python -bb -X dev -m tests.django.manage test tests.django
+    python -bb -X dev -m tests.django.manage test tests.django {posargs}
 
 [testenv:django41]
 commands =
     pip install django~=4.1.0
-    python -bb -X dev -m tests.django.manage test tests.django
+    python -bb -X dev -m tests.django.manage test tests.django {posargs}
 
 [testenv:django42]
 setenv=
     PYTHONWARNDEFAULTENCODING=1
 commands =
     pip install django~=4.2.0
-    python -bb -X dev -m tests.django.manage test tests.django
+    python -bb -X dev -m tests.django.manage test tests.django {posargs}
 
 [testenv:py{37,38,39}-nose]
 deps =

